wasm/wasm.ir/src/org/jetbrains/kotlin/wasm/ir/convertors/BinaryToAst.kt

```diff
@@ -0,0 +1,645 @@
+/*
+ * Copyright 2010-2020 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+@file:Suppress("MemberVisibilityCanBePrivate", "MemberVisibilityCanBePrivate")
+
+package org.jetbrains.kotlin.wasm.ir.convertors
+
+import org.jetbrains.kotlin.wasm.ir.*
+import java.nio.ByteBuffer
+
+
+class WasmBinaryToIR(val b: MyByteReader) {
+    val validVersion = 1u
+
+    val functionTypes: MutableList<WasmFunctionType> = mutableListOf()
+    val structs: MutableList<WasmStructDeclaration> = mutableListOf()
+
+    val importsInOrder: MutableList<WasmNamedModuleField> = mutableListOf()
+    val importedFunctions: MutableList<WasmFunction.Imported> = mutableListOf()
+    val importedMemories: MutableList<WasmMemory> = mutableListOf()
+    val importedTables: MutableList<WasmTable> = mutableListOf()
+    val importedGlobals: MutableList<WasmGlobal> = mutableListOf()
+
+    val definedFunctions: MutableList<WasmFunction.Defined> = mutableListOf()
+    val table: MutableList<WasmTable> = mutableListOf()
+    val memory: MutableList<WasmMemory> = mutableListOf()
+    val globals: MutableList<WasmGlobal> = mutableListOf()
+    val exports: MutableList<WasmExport<*>> = mutableListOf()
+    var startFunction: WasmFunction? = null
+    val elements: MutableList<WasmElement> = mutableListOf()
+    val data: MutableList<WasmData> = mutableListOf()
+    var dataCount: Boolean = false
+
+
+    private fun <T> byIdx(l1: List<T>, l2: List<T>, index: Int): T {
```

Comment:
fun byIdx() = l1..getOrNull(index) ?: l2.getOrNull(index - l1.size) ?: throw "Index not found?"

all code look like java style. toIntOrNull and other checks does not need?

I'm sorry if not in the subject

Reply:
Both variants looks short and clear to me. I'm afraid I don't to see an improvement here.

Reply:
if we try get of max index, then it return null. 

arr[index] == arr.get(index). 

As i remember get throw exceptuion if not found, getOrNull return null..
This funcion will call many times. Point to two times look to code

and becouse my variant is one line without block of logic it can be inline

Reply:
And i vety wait WASM in Kotlin . I need get million objects, get tousends flows and channels connected between. and will get fast data  analyze by 1 read cycle. I need Functional Tensors and data flows. In kotlin not have cuda or other small parralel offload. In llvm+wasm its in box  with compiller optimisation.

Reply:
`getOrNull` is implemented using `get`:
```
public fun <T> List<T>.getOrNull(index: Int): T? {
    return if (index >= 0 && index <= lastIndex) get(index) else null
}
```
So it would not be faster than `get`.
And I don't think it helps with inlining either because there is more code in your version, considering if-expression in getOrNull.

> And i vety wait WASM in Kotlin . I need get million objects, get tousends flows and channels connected between. and will get fast data analyze by 1 read cycle. I need Functional Tensors and data flows. In kotlin not have cuda or other small parralel offload. In llvm+wasm its in box with compiller optimisation.

Nice! Wasm target has a potential to show a good performance, but it is still a long way there.

Reply:
Извини что докопался, но у тебя в примере используется lastIndex, что ограничивает диапазон индекса и get становится безопасно использовать. Смысл разницы в функции getOrNull именно в этом условия. :) За много лет на и глаз уже набит на переполнения... а так я пробно закинул пул реквест для начала с тупым но оч надоедливым багом в корутинах на jvm

Reply:
Мы в любом случае тут падаем при переполнении и это ОК. Или я чего-то еще не замечаю?